src/validationinterface.cpp

```diff
@@ -57,3 +68,176 @@ void UnregisterAllValidationInterfaces() {
+        // Run the notifier on an integer second in the steady clock.
+        auto now = std::chrono::steady_clock::now().time_since_epoch();
+        auto nextFire = std::chrono::duration_cast<std::chrono::seconds>(
+            now + std::chrono::seconds(1));
+        std::this_thread::sleep_until(
+            std::chrono::time_point<std::chrono::steady_clock>(nextFire));
+
+        boost::this_thread::interruption_point();
+
+        auto chainParams = Params();
+
+        //
+        // Collect all the state we require
+        //
+
+        // The common ancestor between the last chain tip we notified and the
+        // current chain tip.
+        const CBlockIndex *pindexFork;
+        // The stack of blocks we will notify as having been connected.
+        // Pushed in reverse, popped in order.
+        std::vector<CachedBlockData> blockStack;
+        // Transactions that have been recently conflicted out of the mempool.
+        std::pair<std::map<CBlockIndex*, std::list<CTransaction>>, uint64_t> recentlyConflicted;
+        // Transactions that have been recently added to the mempool.
+        std::pair<std::vector<CTransaction>, uint64_t> recentlyAdded;
+
+        {
+            LOCK(cs_main);
+
+            // Figure out the path from the last block we notified to the
+            // current chain tip.
+            CBlockIndex *pindex = chainActive.Tip();
+            pindexFork = chainActive.FindFork(pindexLastTip);
+
+            // Fetch recently-conflicted transactions. These will include any
+            // block that has been connected since the last cycle, but we only
+            // notify for the conflicts created by the current active chain.
+            recentlyConflicted = DrainRecentlyConflicted();
+
+            // Iterate backwards over the connected blocks we need to notify.
+            while (pindex && pindex != pindexFork) {
+                // Get the Sprout commitment tree as of the start of this block.
+                SproutMerkleTree oldSproutTree;
+                assert(pcoinsTip->GetSproutAnchorAt(pindex->hashSproutAnchor, oldSproutTree));
+
+                // Get the Sapling commitment tree as of the start of this block.
+                // We can get this from the `hashFinalSaplingRoot` of the last block
+                // However, this is only reliable if the last block was on or after
+                // the Sapling activation height. Otherwise, the last anchor was the
+                // empty root.
+                SaplingMerkleTree oldSaplingTree;
+                if (chainParams.GetConsensus().NetworkUpgradeActive(
+                    pindex->pprev->nHeight, Consensus::UPGRADE_SAPLING)) {
+                    assert(pcoinsTip->GetSaplingAnchorAt(
+                        pindex->pprev->hashFinalSaplingRoot, oldSaplingTree));
+                } else {
+                    assert(pcoinsTip->GetSaplingAnchorAt(SaplingMerkleTree::empty_root(), oldSaplingTree));
+                }
```

Comment:
It might be more elegant to have a method that returns `hashFinalSaplingRoot` or `SaplingMerkleTree::empty_root()` as applicable. Does not block.

Reply:
This will need to be altered anyway as part of #4264 (which re-purposes the block header field); we can address it then.